ipatests/test_integration/test_hsmsupport.py

```diff
@@ -0,0 +1,137 @@
+#
+# Copyright (C) 2018  FreeIPA Contributors see COPYING for license
+#
+"""HSM support for Dogtag PKI
+"""
+from __future__ import absolute_import
+
+import os
+import logging
+
+from ipaplatform.paths import paths
+from ipaplatform.constants import constants
+from ipaplatform.redhat.tasks import selinux_enabled
+from ipatests.test_integration.base import IntegrationTest
+from ipatests.pytest_ipa.integration import tasks
+
+logger = logging.getLogger(__name__)
+
+SELINUX_ENABLED = selinux_enabled()
+SOFTHSM_DIR = '/var/lib/softhsm'
+TOKEN_DIR = os.path.join(SOFTHSM_DIR, 'tokens')
+
+TOKEN_NAME = "softhsm_token"
+TOKEN_PIN = "TokenSecret123"
+TOKEN_SO_PIN = "TokenSOSecret123"
+
+SOFTHSM2_PKI_INI = f"""\
+[DEFAULT]
+pki_hsm_enable=True
+pki_hsm_libfile={paths.LIBSOFTHSM2_SO}
+pki_hsm_modulename=softhsm2
+pki_token_name={TOKEN_NAME}
+"""
+
+SOFTHSM_CMD = [
+    'runuser', '-u', constants.PKI_USER, '--', paths.SOFTHSM2_UTIL,
+]
+
+
+def prepare_softhsm(host, token_name=TOKEN_NAME, token_pin=TOKEN_PIN,
+                    token_so_pin=TOKEN_SO_PIN):
+    """Prepare host for softhsm2
+    """
+    # Workaround for https://pagure.io/dogtagpki/issue/3091, disable
+    # p11-kit-proxy so that Dogtag is able to install SoftHSM2 PKCS#11.
+    host.run_command([
+        'rm', '-f', '/etc/crypto-policies/local.d/nss-p11-kit.config'
+    ])
+    host.run_command(['update-crypto-policies'])
+    # add pkiuser to ods group, so it can create softhsm tokens
+    # see https://bugzilla.redhat.com/show_bug.cgi?id=1625548
+    host.run_command([
+        'usermod', '-G', constants.ODS_GROUP, '-a', constants.PKI_USER
+    ])
+    # change SELinux context from default named_cache_t to pki_tomcat_t to
+    # avoid AVCs for certutil and pkitool
```

Comment:
What are the implications of changing the SELinux context for DNS(SEC), which also uses softhsm (IIRC)?

Reply:
This is a hack to have some very basic level of testing. There is no proper solution yet. Both Dogtag and SoftHSM needs to improve to officially support SoftHSM with Dogtag.

IPA's DNSSEC integration uses a custom softhsm config file and a different storage location for the tokens.